src/Microsoft.TestPlatform.CrossPlatEngine/TestSession/TestSessionPool.cs

```diff
@@ -61,7 +67,7 @@ public static TestSessionPool Instance
         /// <param name="proxyManager">The proxy manager object.</param>
         /// 
         /// <returns>True if the operation succeeded, false otherwise.</returns>
-        public bool AddSession(
+        public virtual bool AddSession(
             TestSessionInfo testSessionInfo,
             ProxyTestSessionManager proxyManager)
```

Comment:
I'm wondering why is there a separation between the session manager and the session pool?

Reply:
Well, the session manager (`ProxyTestSessionManager`) is the "active" component that manages all the resources associated with a single session identified by its `TestSessionInfo` object, which in this case are the testhosts that are being spawned in order to fulfill the incoming discovery/run requests. This is mostly analogous to the way `ProxyDiscoveryManager` and `ProxyExecutionManager` work.

The `TestSessionPool` is basically a mapping between the `TestSessionInfo` and its corresponding `ProxyTestSessionManager` and allows us to support multiple simultaneous test sessions. All those different sessions are managed by their own proxy, and all those proxies are managed by the `TestSessionPool`.